Harmonizing serialization of Parameters and Extensions. (#761)
moq-wg/moq-transport: draft-ietf-moq-transport.md

`A.Group < B.Group || (A.Group == B.Group && A.Object < B.Object)`

### Key-Value-Pair Structure

Key-Value-Pair is a flexible structure designed to carry key/value
pairs in which the key is a variable length integer and the value
is either a variable length integer or a byte field of arbitrary
length.

Key-Value-Pair is used in both the data plane and control plane, but

> is optimized for use in the data plane.

Reviewer: I think this may be misleading to say its optimized for data plane. I wonder if we should remove that statement for now and resolve https://github.com/moq-wg/moq-transport/issues/794

Reviewer 2: I view the statement as the intention of the wg moving forward and we should leave it in place in this PR.

Reviewer 3: I thought this would help address Mo point that people reading it later would not forget that from a design point of view what we trying to do.

~~~
Key-Value-Pair {
  Type (i),
  [Length (i),]
  Value (..)
}
~~~
{: #moq-key-value-pair format title="MOQT Key-Value-Pair"}

* Type: an unsigned integer, encoded as a varint, identifying the
  type of the value and also the subsequent serialization.

> * Length: Only present when Type is odd. Species the length of the

Reviewer: ```suggestion
* Length: Only present when Type is odd. Specifies the length of the
```

> * Length: Only present when Type is odd. Specifies the length of the

Reviewer 3: I don't like the even / odd way to split this. I would rather see us defined a range of numbers that have no length and range that has a length. This allows us to adjust those two ranges to not be the same size. We have such a limited space for short codes that I don't want to waste them.

Reviewer 3: My suggestions would 0 to 10, no length, 53 to 63 has length, and 11 to 52 reserved and we can sort out how to allocate that as we get spec gets closer to RFC and we understand the usage patters better.

Reviewer 2: We're not making a decision about even/odd in this PR, we are deciding to use the *same format* to serialize Parameters and Extensions.  Please add your comments to #794.

|------|---------------------------|
| 0x4  | Duplicate Track Alias     |
|------|---------------------------|

> | 0x5  | Parameter Value Mismatch |

Reviewer 4: I like Parameter Length Mismatch more, even with this change.

Reviewer 4: Looking at the text below again, I think there are two potential issues.  One is that you try to read a varint or variable length field and run out of bytes.  The other is that the value read doesn't match the valid values for the parameter or extension.

I think the first is a length issue, though calling it a mismatch might be misleading?

The latter I might call "Invalid Parameter Value" or "Parameter Decoding Error" or ?

Author: Can we find a name that is broad enough to capture the 3 error conditions under which we night want to throw:
1. The varint defining the type is not well formatted
2. The length declared does not match the payload size
3. The extension is not well formatted for its type (even or odd)
4. The parameters have not been serialized in ascending order. 

How about "Parameter Formatting Error" ? 

or do we want to throw different errors for the different conditions?

Reviewer 2: Who said #273

Reviewer 4: I'm fine with "Parameter Formatting Error" or "Parameter Parsing Error"?

Author: Changed error text to "parameter Formatting Error" with commit https://github.com/moq-wg/moq-transport/pull/761/commits/8e8257db9adc4d51d83b07e15191181d19bdf7da

|-------|-----------------------------------------------------|
| ID    | Messages                                            |
|------:|:----------------------------------------------------|

> | 0x50  | CLIENT_SETUP ({{message-setup}})                    |

Reviewer 2: I asked Will to renumber the SETUP messages so you can write an implementation that supports both draft <= 10 and > 10.  If we kept the code point the same but changed the format of the message that negotiates version, I don't know how to do that.

> | 0x50  | CLIENT_SETUP ({{message-setup}})                    |

Reviewer 4: Thanks!

> | 0x50  | CLIENT_SETUP ({{message-setup}})                    |

Reviewer 3: I do not think we should be doing this type of backwards compatibility at this point. If you want to do something that support 10 and 11, use the ALPN to sort it out. I'd rather just note things and breaking changes.  This just makes the draft more confusing to people reading it and adds to cruft code.

## Parameters {#params}

Some messages include a Parameters field that encode optional message
elements. They contain a type, an optional length, and a value.

Senders MUST NOT repeat the same parameter type in a message. Receivers
SHOULD check that there are no duplicate parameters and close the session
as a 'Protocol Violation' if found.

Receivers ignore unrecognized parameters.

Parameters MUST be serialized in ascending Parameter Type order.

The format of Parameters is as follows:

~~~
Parameter {
  Parameter Type (i),
  [Parameter Value (i)]
  [Parameter Length (i),

>    Parameter Value (..)]

Reviewer 2: See the way that I changed Extension definition to not require multi-line [] optional notation.

Author: Resolved with https://github.com/moq-wg/moq-transport/pull/761/commits/75498823b4e13e3b97b27ba1ca21a43f09ad6aa7

## Parameters {#params}

Some messages include a Parameters field that encode optional message
elements. They contain a type, an optional length, and a value.

Senders MUST NOT repeat the same parameter type in a message. Receivers
SHOULD check that there are no duplicate parameters and close the session
as a 'Protocol Violation' if found.

Receivers ignore unrecognized parameters.

> Parameters MUST be serialized in ascending Parameter Type order.

Reviewer 5: Or else what? (I supect you want to say "the receiver MUST terminate the session with error code 'Parameter Value Mismatch'.")

Reviewer 4: I actually think with this encoding, a MUST isn't necessary because you can't do anything else.

```suggestion
Parameters are serialized in ascending Parameter Type order using a delta encoding.
```

Reviewer: I am not really sure if we need the serializaiton requirement and/or the delta encoding .. I haven;t thought, but I am not sure how would this work when we register things with IANA where the assigned number space may not fit with the proposed way of encoding.

Reviewer 4: It's easier to validate a parameter if it's all in one place, rather than one instance at the beginning and another at the end, based on my experience with HTTP headers.  HTTP headers are strings, so it's not as trivial to enforce ordering, but parameters have assigned varint codepoints, so ordering is obvious.

Reviewer 2: Does the ascending order requirement break what @fluffy is trying to do with authenticated extensions?

Author: Removed the requirement to serialize in ascending type order with commit https://github.com/moq-wg/moq-transport/pull/761/commits/17db4c969a0a46a309ca6b960d1b207da9030ea8.

It is not needed since we removed the delta encoding. Parameters can now be serialized in any type order.

> Parameters MUST be serialized in ascending Parameter Type order.

Reviewer 3: Depending on how integrrity works, you might not want this. If some of them are end to end protected and some are not, you can probably not do this. I don't see the reason to do this.

Reviewer 3: I'm very concerned with delta encoding for this. That would mean a relay that add a extension would have to rewrite all the following extensions and seem like it probably won't save much space.

Author: Commits https://github.com/moq-wg/moq-transport/pull/761/commits/8e8257db9adc4d51d83b07e15191181d19bdf7da and https://github.com/moq-wg/moq-transport/pull/761/commits/17db4c969a0a46a309ca6b960d1b207da9030ea8  removed delta encoding and requirement for serializing in asecnding order.

The format of Parameters is as follows:

~~~
Parameter {
  Parameter Type (i),
  [Parameter Value (i)]
  [Parameter Length (i),
   Parameter Value (..)]
}
~~~
{: #moq-param format title="MOQT Parameter"}

> Parameter Type is an unsigned integer that indicates the semantic meaning of the

Reviewer: should we say variable length integer instead?

parameter and also the subsequent serialization. Parameter Type is delta encoded.
The first Parameter Type in the message is encoded with an absolute value and all
subsequent Parameter Types are encoded as a delta to the prior Parameter Type.
Setup message parameters use a namespace that is constant across all MoQ Transport
versions. All other messages use a version-specific namespace. For example, the
integer '1' can refer to different parameters for Setup messages and for all other
message types. SETUP message parameter types are defined in {{setup-params}}.
Version-specific parameter types are defined in {{version-specific-params}}.

Parameter Values: even Parameter Types are followed by a single varint encoded
value. Odd Parameter Types are followed by the Parameter Value length in bytes

> followed by the Parameter Value.

Reviewer: Even/Odd idea is cool , but I wonder this is overly optimizing somehting that is in control parameters.

> followed by the Parameter Value.

Reviewer 4: It's nice to have a single parser for both, though.

before sending any data for that Object, taking into account priorities,
congestion control, and any other relevant information.

#### AUTHORIZATION INFO {#authorization-info}

AUTHORIZATION INFO parameter (Parameter Type 0x03) identifies a track's
authorization information in a SUBSCRIBE, SUBSCRIBE_ANNOUNCES, ANNOUNCE
or FETCH message. This parameter is populated for cases where the authorization

> is required at the track level. The value is an ASCII string.

Reviewer: s/ASCII String/ array of bytes

Reviewer 2: +1 I thought we removed ASCII string already, did it find it's way back.

Author: I rebased with https://github.com/moq-wg/moq-transport/pull/761/commits/75498823b4e13e3b97b27ba1ca21a43f09ad6aa7 and that restored the removal.

> is required at the track level. The value is an ASCII string.

Reviewer: ```suggestion
is required at the track level. The value is array of bytes.
```

Author: The value is defined by the serialization, so no need to repeat here.

before sending any data for that Object, taking into account priorities,
congestion control, and any other relevant information.

> #### AUTHORIZATION INFO {#authorization-info}

Reviewer 3: I don't think section should be in the same PR

Reviewer 2: AUTH INFO has to change it's code point in this PR (from 2 to 3).  Otherwise it's just moving text.  Do you want to keep AUTH INFO before delivery timeout?  Nothing else changed.
